src/s3path.jl

```diff
@@ -0,0 +1,278 @@
+    end
+end
+
+FilePathsBase.ispathtype(::Type{S3Path}, str::AbstractString) = startswith(str, "s3://")
+FilePathsBase.exists(fp::S3Path) = s3_exists(fp.config, fp.bucket, fp.key)
+Base.isfile(fp::S3Path) = !fp.isdirectory && exists(fp)
+function Base.isdir(fp::S3Path)
+    if isempty(fp.segments)
+        key = ""
+    elseif fp.isdirectory
+        key = fp.key
+    else
+        return false
+    end
+
+    objects = s3_list_objects(fp.config, fp.bucket, key; max_items=1)
+    return !isempty(objects)
+end
+
+function Base.stat(fp::S3Path)
+    # Currently AWSS3 would require a s3_get_acl call to fetch
+    # ownership and permission settings
+    m = Mode(user=(READ + WRITE), group=(READ + WRITE), other=(READ + WRITE))
+    u = FilePathsBase.User()
+    g = FilePathsBase.Group()
+    blksize = 4096
+    blocks = 0
+    s = 0
+    last_modified = DateTime(0)
+
+    if exists(fp)
+        resp = s3_get_meta(fp.config, fp.bucket, fp.key)
+        # Example: "Thu, 03 Jan 2019 21:09:17 GMT"
+        last_modified = DateTime(
+            resp["Last-Modified"][1:end-4],
+            dateformat"e, d u Y H:M:S",
+        )
+        s = parse(Int, resp["Content-Length"])
+        blocks = ceil(Int, s / 4096)
+    end
+
+    return Status(0, 0, m, 0, u, g, 0, s, blksize, blocks, last_modified, last_modified)
+end
+
+# Need API for accessing object ACL permissions for this to work
+FilePathsBase.isexecutable(fp::S3Path) = false
+Base.isreadable(fp::S3Path) = true
+Base.iswritable(fp::S3Path) = true
+Base.ismount(fp::S3Path) = false
+
+function Base.mkdir(fp::S3Path; recursive=false, exist_ok=false)
+    fp.isdirectory || throw(ArgumentError("S3Path folders must end with '/': $fp"))
+
+    if exists(fp)
+        !exist_ok && error("$fp already exists.")
+    else
+        if hasparent(fp) && !exists(parent(fp))
+            if recursive
+                mkdir(parent(fp); recursive=recursive, exist_ok=exist_ok)
+            else
+                error(
+                    "The parent of $fp does not exist. " *
+                    "Pass `recursive=true` to create it."
+                )
+            end
+        end
+
+        write(fp, "")
+    end
+
+    return fp
+end
+
+function Base.rm(fp::S3Path; recursive=false, kwargs...)
+    if isdir(fp)
+        files = readpath(fp)
+
+        if recursive
+            for f in files
+                rm(f; recursive=recursive, kwargs...)
+            end
+        elseif length(files) > 0
+            error("S3 path $object is not empty. Use `recursive=true` to delete.")
+        end
+    end
+
+    @debug "delete: $fp"
+    s3_delete(fp.config, fp.bucket, fp.key)
+end
+
+# We need to special case sync with S3Paths because of how directories
+# are handled again.
+function FilePathsBase.sync(src::AbstractPath, dst::S3Path; delete=false)
+    # Create an index of all of the source files
+    index = Dict(Tuple(setdiff(p.segments, src.segments)) => p for p in walkpath(src))
```

Comment:
Does this always work?
```julia
julia> setdiff(["a", "b", "c"], ["b", "a", "d"])
1-element Array{String,1}:
 "c"
```

Reply:
Yes? We can trust that `src.segments` is always a direct subset of `p.segments` based on how `walkpath` works. Is there something else in your example that I'm supposed to be noticing?

Reply:
Hmm, I see. Seems to be a slight abuse of `setdiff` but this should be ok